src/Microsoft.SqlTools.ServiceLayer/DacFx/GetObjectsFromTSqlModelOperation.cs

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.SqlTools.ServiceLayer.DacFx.Contracts;
+using Microsoft.SqlTools.Utility;
+using Microsoft.SqlServer.Dac.Model;
+
+namespace Microsoft.SqlTools.ServiceLayer.DacFx
+{
+    /// <summary>
+    /// Class to represent request to get objects from model
+    /// </summary>
+    class GetObjectsFromTSqlModelOperation
+    {
+        public TSqlModel Model;
+        public GetObjectsFromTSqlModelParams Parameters { get; }
+
+        public GetObjectsFromTSqlModelOperation(GetObjectsFromTSqlModelParams parameters, TSqlModel model)
+        {
+            Validate.IsNotNull("parameters", parameters);
+            this.Parameters = parameters;
+            this.Model = model;
+        }
+
+        /// <summary>
+        /// Get user defined top level type objects from model
+        /// </summary>
+        public TSqlObjectInfo[] GetObjectsFromTSqlModel()
+        {
+            try
+            {
+                List<ModelTypeClass> filters = new List<ModelTypeClass>();
+                foreach (var type in Parameters.ObjectTypes)
+                {
+                    filters.Add(MapType(type));
+                }
+                var objects = Model.GetObjects(DacQueryScopes.UserDefined, filters.ToArray()).ToList();
+
+                return objects.Select(o => new TSqlObjectInfo
+                {
+                    Name = o.Name.ToString(),
+                    ObjectType = o.ObjectType.Name,
+                    ReferencedObjects = o.GetReferencedRelationshipInstances().ToList().Select(r => new TSqlObjectRelationship
+                    {
+                        Name = r.ObjectName.ToString(),
+                        ObjectType = r.Object.ObjectType.Name,
+                        RelationshipType = r.Relationship.Type.ToString(),
+                        FromObjectName = r.FromObject.Name.ToString(),
+                        FromObjectType = r.FromObject.ObjectType.Name
+                    }).ToArray()
+                }).ToArray();
+            }
+            catch (Exception ex)
+            {
+                Logger.Write(TraceEventType.Information, $"Failed to get user defined objects from model. Error: {ex.Message}");
+                throw;
+            }
+        }
+
+        public static ModelTypeClass MapType(string type)
+        {
+            switch (type)
+            {
+                case "table": return ModelSchema.Table;
+                case "view": return ModelSchema.View;
+                default:
+                    throw new ArgumentException($@"Unsupported data source type ""{type}""",
```

Comment:
the `@` shouldn't be needed for string interpolation. Do the double quotes need to be escaped? Single quotes should be ok to use too.